File: 0219. Contains Duplicate II/0219.py
# time complexity O(n)
# space complexity O(min(n, k))
class Solution(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        # solution using dictionary (hashmap)
        check_dict = {} # to store the last seen index of each number

        for i in range(0, len(nums)):
            if nums[i] in check_dict and i - check_dict.get(nums[i]) <= k:
                return True
            check_dict[nums[i]]=i # add the element. if exist, update.

        return False
        

File: 0219. Contains Duplicate II/0219.java
// time complexity O(n)
// space complexity O(1)

class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        // hashmap solution
        Map<Integer, Integer> check = new HashMap<>();

        for(int i=0; i<nums.length; i++){
            if(check.containsKey(nums[i]) && 
            Math.abs(check.get(nums[i]) - i) <= k) return true;
            else check.replace(nums[i], i);

            check.put(nums[i], i);
        }

        return false;
    }
}